### PR TITLE
LIGO Integration Tests

### LINKED ISSUE #4: Burn rate following "is_address_implicit" not working as intended
Burn rate following "is_address_implicit" is actually not working as intended.

There is no burn rate on "tz1..." addresses, while it should be the case for "kt1..." addresses.

This is due to a counter-condition that is badly declared.

Implementing a "not" clause should solve the issue.

Preparing a pull request right now...

### PR DESCRIPTION
**Pull request by ghost**
* Adding LIGO Integration Tests

* Fixing is_address_implicit() bug

* Misc. Updates

### CODE REVIEW

File: test/helpers/assert.mligo [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    let _expected = Test.eval expected in
+    let () = match res with
+        | Fail (Rejected (actual,_)) -> assert (actual = _expected)
+        | Fail (Balance_too_low err) -> failwith "Contract failed: Balance too low"
+        | Fail (Other s) -> failwith s
+        | Success _ -> failwith "Transaction should fail"
+    in
+    Test.log("OK", expected)
+
+let assert_transfer_account(ant_addr, sender, recipient, init_token_balance, tsfr_amount : (ANTI.parameter, ANTI.storage) typed_address * address * address * nat * nat) : unit =
```
Comment (ziwot) [inline test/helpers/assert.mligo]: I think it should be more generic : assert_balance(taddr, address, amount)
An assert function should not contain the values it asserts, they must be passed as param

File: test/functions.mligo [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    let () = Test.set_source from_ in
+    let approve_requests = ({spender=spender_; value=amount_} : ANTI.approve) in
+    Test.transfer_to_contract contr (Approve approve_requests) 0tez
+
+let transfer(contr, from_, to_, amount_ : ANTI.parameter contract * address * address * nat) =
+    let () = Test.set_source from_ in
+    let transfer_requests = ({address_from=from_; address_to=to_; value=amount_} : ANTI.transfer) in
+    Test.transfer_to_contract contr (Transfer transfer_requests) 0tez
+
+let approved_transfer(contr, from_, to_, amount_ : ANTI.parameter contract * address * address * nat) =
```
Comment (ziwot) [inline test/functions.mligo]: identical to the function above
I think this file should be named anti_helper, this way, in the tests, it would be called like this: `Anti_helper.transfer(...` instead of `FUNC.transfer(...`

File: test/helpers/assert.mligo [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    let () = assert(rtr_bal_brn = abs((tsfr_amount * 7n) / 100)) in
+    let () = assert(rtr_bal_res = abs((tsfr_amount * 1n) / 100)) in
+
+    let () = Test.log("OK, Sender :", rtr_bal_sender) in
+    let () = Test.log("OK, Recipient :", rtr_bal_recipient) in
+    let () = Test.log("OK, Burn address :", rtr_bal_brn) in
+    let () = Test.log("OK, Reserve address :", rtr_bal_res) in
+    ()
+
+let assert_transfer_contract(ant_addr, sender, recipient, init_token_balance, tsfr_amount : (ANTI.parameter, ANTI.storage) typed_address * address * address * nat * nat) : unit =
```
Comment (ziwot) [inline test/helpers/assert.mligo]: identical to the function above, but with different asserted values inside

File: test/helpers/errors.mligo [hunk 4]
```diff
@@ -0,0 +1,4 @@
+let err_rejected : string = "Rejected"
```
Comment (ziwot) [inline test/helpers/errors.mligo]: we are allowed to edit the contract under test, you can add an errors module there and just test that contract calls returns expected errors `Anti.Errors.not_enough_balance` as an example
if we duplicate errors, we have to maintain 2 identical files, and it's likely that some time they will diverge, like here `err_rejected` doesn't exists

File: test/main.mligo [hunk 5]
```diff
@@ -0,0 +1,156 @@
+#import "../src/anti.mligo" "ANTI"
+#import "./functions.mligo" "FUNC"
+#import "./contracts/CALLBACK.mligo" "CALLBACK"
+#import "./helpers/errors.mligo" "ERROR"
+#import "./helpers/assert.mligo" "ASSERT"
+
+let test =
```
Comment (ziwot) [inline test/main.mligo]: this global wrapper is'nt needed, it will appear like there is only one test whereas if all the tests are at the top-level, they will appear clearly int the terminal

File: test/main.mligo [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        let () = assert(retrieved_balance = init_token_balance) in
+        Test.log("OK, Alice :", retrieved_balance)
+    in
+    
+    (* Testing GetBalance Callback for Alice *)
+    let test_get_balance_callback = 
+        let () = Test.log("--> test_get_balance_callback : Testing GetBalance Callback for Alice") in
+        let balance_of_request = ({
+            owner = alice;
+            callback = cbk_addr;
```
Comment (ziwot) [inline test/main.mligo]: should be of type `nat contract`

File: test/main.mligo [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    
+    (* KO : Transferring 300 from Bob to Alice *)    
+    let test_transfer_with_no_balance =
+        let () = Test.log("--> test_transfer_with_no_balance : Transferring 300 from Bob to Alice") in
+        let tsfr_amount = 300n in
+        let result : test_exec_result = FUNC.transfer(ant_ctr, bob, alice, tsfr_amount) in
+        ASSERT.string_failure result ERROR.err_not_enough_balance
+    in
+
+    let (ant_ctr, ant_addr, cbk_addr, cbk_ctr, alice, bob, james) = FUNC.bootstrap init_token_supply init_token_balance in
```
Comment (ziwot) [inline test/main.mligo]: I think each test should follow an arrange/act/assert or given/when/then pattern
for that, each test should bootstrap its environment (arrange)
because otherwise we have this pattern :
arrange, act, assert, act, assert,act.assert, arrange making tests hard to follow or debug

File: test/main.mligo [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    (* OK : Setting Allowance for James from Alice *)
+    let test_allowance_set =
+        let () = Test.log("--> test_allowance_set : Setting 300n Allowance for James from Alice") in
+        let allwn_amount = 300n in
+        let result : test_exec_result = FUNC.approve(ant_ctr, alice, james, allwn_amount) in
+        ASSERT.tx_success result
```
Comment (ziwot) [inline test/main.mligo]: should also test the storage (everywhere)

File: test/main.mligo [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    let test_allowance_calllback =
+        let () = Test.log("--> test_allowance_calllback : Testing GetAllowance Callback for James from Alice") in
+        let allowance_request = ({
+            request =   { owner = alice; spender = james };
+            callback = cbk_addr;
+        } : ANTI.getAllowance) in
+        let result_cbk : test_exec_result = Test.transfer_to_contract ant_ctr (GetAllowance allowance_request) 0tez in
+        let () = ASSERT.tx_success result_cbk in
+        let cbk_storage = Test.get_storage cbk_ctr in
+        let () = assert(cbk_storage = 300n) in
```
Comment (ziwot) [inline test/main.mligo]: here is why I think each test should be independent, when reading it, we need to look at the test above to understand where does the 300n value comes from

File: test/main.mligo [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        let _burn_address : address = ("tz1burnburnburnburnburnburnburjAYjjX" : address) in
+        let burn_balance : nat = FUNC.get_balance_from_storage(ant_addr, _burn_address) in
+        let supply_request = ({
+            request =  ();
+            callback = cbk_addr;
+        } : ANTI.getTotalSupply) in
+        let result = Test.transfer_to_contract ant_ctr (GetTotalSupply supply_request) 0tez in
+        let () = ASSERT.tx_success result in
+        let cbk_storage = Test.get_storage cbk_ctr in
+        let () = assert(cbk_storage = abs(init_token_supply - burn_balance)) in
```
Comment (ziwot) [inline test/main.mligo]: if it is only the callback that is tested, it should only assert that the received value is same at the one in the storage

File: Makefile [hunk 11]
```diff
@@ -1,26 +1,36 @@
 ifndef LIGO
-LIGO=docker run --rm -v "$(PWD)":"$(PWD)" -w "$(PWD)" ligolang/ligo:next
+LIGO=docker run --platform linux/amd64 --rm -v "$(PWD)":"$(PWD)" -w "$(PWD)" ligolang/ligo:0.43.0
 endif
 
-json=--michelson-format json
+PROTOCOL_OPT=--protocol ithaca
```
Comment (ziwot) [inline Makefile]: not sure this is needed, we add this flag when testing specific features https://ligolang.org/docs/protocol/ithaca

File: test/helpers/errors.mligo (hunk 4 above)
Comment (ghost) [reply test/helpers/errors.mligo]: "We are allowed to edit the contract under test"
How would you proceed to do that ?
Isn't adding an error module there is the same as adding an error module in the Testing file ?

File: test/main.mligo (hunk 6 above)
Comment (ghost) [reply test/main.mligo]: The callback actually renders an error on my side if I use the `nat contract` type
I don't find any use of this type on the docs
Do you have an example on hand ?

File: test/main.mligo (hunk 7 above)
Comment (ghost) [reply test/main.mligo]: I think this is already the case in the current test
The concurrent two tests are adjacent so that we can test a specific case : sending 100 to Bob, and trying a 300 transfer with insufficient balance back from Bob.
If we bootstrap the environment again; Bob won't have any tokens (which is the following test line 82 with James)

File: test/main.mligo (hunk 9 above)
Comment (ghost) [reply test/main.mligo]: I don't think these two tests can be independent.
The GetAllowance needs the SetAllowance to work.

Do you wanna group them both in one single (bigger) function ?

File: test/main.mligo (hunk 5 above)
Comment (ghost) [reply test/main.mligo]: The global wrapper is temporary, as it will host the other functions once they have all been reviewed and cut into small files

File: test/functions.mligo (hunk 2 above)
Comment (ghost) [reply test/functions.mligo]: Done.

File: test/helpers/assert.mligo (hunk 1 above)
Comment (ghost) [reply test/helpers/assert.mligo]: The current params are (contract, sender, recipient, before_amount, transfer_amount)
The only coded values inside the Assert are the Burn/Reserve percentages
What values are not good for you ?

File: test/helpers/assert.mligo (hunk 3 above)
Comment (ghost) [reply test/helpers/assert.mligo]: Same as above

File: test/main.mligo (hunk 10 above)
Comment (ghost) [reply test/main.mligo]: Done.

File: test/main.mligo (hunk 9 above)
Comment (ziwot) [reply test/main.mligo]: no, you can set up desired initial storage and just test one entrypoint

File: test/main.mligo (hunk 5 above)
Comment (ziwot) [reply test/main.mligo]: not sure I understand, as you want

File: test/helpers/assert.mligo (hunk 1 above)
Comment (ziwot) [reply test/helpers/assert.mligo]: Sorry, I though the values were something else, I guess the problem is in the contract, `1` and `7` should be constants

File: test/helpers/assert.mligo (hunk 3 above)
Comment (ziwot) [reply test/helpers/assert.mligo]: I could be more readable with a simple generic assert_balance(taddr, address, amount), leaving each test giving the expected values, having all the information at one place, in the test

File: test/main.mligo (hunk 6 above)
Comment (ziwot) [reply test/main.mligo]: I don't know maybe there is a problem in the contract, but it the type appearing here https://github.com/smart-chain-fr/mechanical-token/blob/main/src/anti.mligo#L25

File: test/helpers/errors.mligo (hunk 4 above)
Comment (ziwot) [reply test/helpers/errors.mligo]: the purpose is to avoid duplication

**Review by ghost — COMMENTED** [review]
Submitted discussed changes on 27th June 2022

**Review by frankhillard — APPROVED** [review]
